pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/ComplexTypeTransformer.java

```diff
@@ -0,0 +1,229 @@
+    }
+    return record;
+  }
+
+  private void unnestCollection(GenericRow record, String column, List<GenericRow> list) {
+    Object value = record.removeValue(column);
+    if (value == null) {
+      // use the record itself
+      list.add(record);
+    } else if (value instanceof Collection) {
+      if (((Collection) value).isEmpty()) {
+        // use the record itself
+        list.add(record);
+      } else {
+        for (Object obj : (Collection) value) {
+          GenericRow copy = flattenCollectionItem(record, obj, column);
+          list.add(copy);
+        }
+      }
+    } else if (value.getClass().isArray()) {
+      if (((Object[]) value).length == 0) {
+        // use the record itself
+        list.add(record);
+      } else {
+        for (Object obj : (Object[]) value) {
+          GenericRow copy = flattenCollectionItem(record, obj, column);
+          list.add(copy);
+        }
+      }
+    }
+  }
+
+  private GenericRow flattenCollectionItem(GenericRow record, Object obj, String column) {
+    GenericRow copy = record.copy();
+    if (obj instanceof Map) {
+      Map<String, Object> map = (Map<String, Object>) obj;
+      for (Map.Entry<String, Object> entry : new HashSet<>(map.entrySet())) {
+        String flattenName = concat(column, entry.getKey());
+        copy.putValue(flattenName, entry.getValue());
+      }
+    } else {
+      copy.putValue(column, obj);
+    }
+    return copy;
+  }
+
+  @VisibleForTesting
+  protected GenericRow flattenMap(GenericRow record, Collection<String> columns) {
+    for (String column : columns) {
+      if (record.getValue(column) instanceof Map) {
+        Map<String, Object> map = (Map) record.removeValue(column);
+        List<String> mapColumns = new ArrayList<>();
+        for (Map.Entry<String, Object> entry : new HashSet<>(map.entrySet())) {
+          String flattenName = concat(column, entry.getKey());
+          record.putValue(flattenName, entry.getValue());
+          mapColumns.add(flattenName);
+        }
+        record = flattenMap(record, mapColumns);
+      } else if (record.getValue(column) instanceof Collection && _collectionsToUnnest.contains(column)) {
+        for (Object inner : (Collection) record.getValue(column)) {
+          if (inner instanceof Map) {
+            Map<String, Object> innerMap = (Map<String, Object>) inner;
+            flattenMap(column, innerMap, new HashSet<>(innerMap.keySet()));
+          }
+        }
+      }
+    }
+    return record;
+  }
+
+  private void flattenMap(String path, Map<String, Object> map, Collection<String> fields) {
+    for (String field : fields) {
+      if (map.get(field) instanceof Map) {
+        Map<String, Object> innerMap = (Map<String, Object>) map.remove(field);
+        List<String> innerMapFields = new ArrayList<>();
+        for (Map.Entry<String, Object> innerEntry : new HashSet<>(innerMap.entrySet())) {
+          map.put(concat(field, innerEntry.getKey()), innerEntry.getValue());
```

Comment:
Cache `concat(field, innerEntry.getKey()`, and check the value before putting it to the list for the next iteration